[AOTInductor] Proxy Executor for Extern Fallback kernels

SherlockNoMad opened the pull request:
Summary:
This is a prototype for running extern fallback kernels with a host side proxy executor.

Sample of generated cpp wrapper call:
```
        at::Tensor buf0;  // output buffer
        void* tensor_args_var_0[] = {&arg0_1, &arg0_1, &arg1_1, &arg0_1, &arg1_1, &buf0};
        int64_t int_args_var_1[] = {81, 81, 7, 7, 7, 81};
        proxy_executor->call_function("buf0", int_args_var_1, tensor_args_var_0);
```

- In my current implementation, proxy executor interprets the raw pointers according to the ops schema.
This assumes that custom op MUST have a valid schema registered to Dispatcher. (I would like to validate this assumption)
- I am using callboxed() API of the custom kernels. This is inevitable, as we wish to have a single call_function API for all possible custom kernels.

- These are all the input argument types I have support so far.
       union Argument {
         # Bool value does not matter
         1: bool asNone;
         2: TensorArgument asTensor;
         3: list<TensorArgument> asTensors;
         5: i64 asInt;
         7: list<i64> asInts;
         8: double asFloat;
         9: list<double> asFloats;
         10: string asString;
         10.5: list<string> asStrings;
         11: SymIntArgument asSymInt;
         12: list<SymIntArgument> asSymInts;
         13: ScalarType asScalarType;
         14: MemoryFormat asMemoryFormat;
         15: Layout asLayout;
         16: Device asDevice;
         17: bool asBool;
         18: list<bool> asBools;
       }

- Need a policy for handling unpopulated argument with default values. Here are the options, and it has BC  implications.
1. requires exported fx graph to explicitly populate default values, if users doesn't specify.
2. requires cpp wrapper to explicitly populate default values, if fx graph doesn't specify.
3. Proxy executor look up from opSchema for default values.

Test Plan:
frontend:
buck2 run mode/dev-sand mode/inplace -c fbcode.enable_gpu_sections=True sigmoid/frontend:export_main

test:
 buck2 run mode/dev-sand //deeplearning/aot_inductor/test:test_custom_ops

backend:
buck2 run mode/dev-nosan //deeplearning/aot_inductor/fb:main

Differential Revision: D48006843



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov


ShipIt is confused by this PR. Since the original diff was reverted and the PR was never merged, adding:
@diff-train-skip-merge

ezyang commented on torch/csrc/inductor/proxy_executor.h:
```diff
@@ -0,0 +1,24 @@
+namespace torch {
+namespace aot_inductor {
+
+class TORCH_API ProxyExecutor : public torch::CustomClassHolder {
```
BTW, there isn't actually any good reason for this to be a CustomClassHolder, right? All you really need is it to be intrusively refcounted. (Do you even need refcounting on it?)

SherlockNoMad replied:
Yes, I need it to be intrusively refcounted, as it's used in FbAOTInductorModel's def_pickle. 

I wonder how to make this intrusively refcounted, but via CustomClassHolder?

ezyang replied:
Why do we pickle the proxy executor 🤔

ezyang commented on torch/_inductor/__init__.py:
```diff
@@ -49,7 +49,13 @@ def aot_compile(
+    # AOTInductor returns result as a string, not callable
+    # Maybe this check is not neded?
+    if callable(result):
+        result = result()
```
It would be very nice to not have this check

ezyang commented on torch/_inductor/codegen/aot_inductor_interface.cpp:
```diff
@@ -69,8 +71,11 @@ AOTInductorError AOTInductorModelContainerRun(
 
   auto stream = reinterpret_cast<cudaStream_t>(stream_handle);
+
+  torch::aot_inductor::ProxyExecutor* proxy_executor = reinterpret_cast<torch::aot_inductor::ProxyExecutor*>(proxy_executor_handle);
```
@desertfire this will need to be ABI'ified

ezyang commented on torch/_inductor/ir.py:
```diff
@@ -3690,8 +3701,42 @@ def find_device(tensor_args, example_output):
             return devices[0]
         return None
 
+    def export_extern_kernel_node(self):
```
There should be a design comment somewhere describing the overall design of how the proxy executor works (think the high level design doc we hashed through before you started this). Then, this function should cross-reference with that design doc. If you like, you can make the gdoc public and link it here.

SherlockNoMad replied:
Done.

ezyang commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -914,6 +916,7 @@ def __init__(self):
         self.cuda = False
         self.supports_intermediate_hooks = False
         self.outputs_need_copy = set()
+        self.kernel_callsite_id = count()
```
nice, thanks

ezyang commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -1295,6 +1407,43 @@ def generate_extern_kernel_alloc_and_find_schema_if_needed(
+            f"{len(int_call_args)}, "
+            f"{int_args_var}, "
+            f"{len(tensor_call_args)}, "
+            f"{tensor_args_var});"
```
@desertfire this will need to be ABI'ified

ezyang commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -1272,6 +1276,83 @@ def make_buffer_allocation(self, buffer):
+            elif isinstance(return_type, torch.ListType) and isinstance(
+                return_type.getElementType(), torch.TensorType
+            ):
+                # TODO: handle tensor list return type
```
Wouldn't it be better to raise NotImplementedError here, rather than silently fail to read out the outputs entirely?

SherlockNoMad replied:
fixed.

ezyang commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -1272,6 +1276,83 @@ def make_buffer_allocation(self, buffer):
+                        assert all(
+                            is_int_type
+                        ), "AOTInductor only supports int scalars of the same type"
+                        new_int_args.extend([str(a) for a in arg])
```
A compromise might be to somehow have a single function that extracts the dynamic arguments that can be shared server-client.

SherlockNoMad replied:
I added an list for static types. Should be pretty close to exhaustive, but might need update as we support more custom ops.

ezyang reviewed (approved):
Some basic code understandability questions.

SherlockNoMad commented on torch/_inductor/codecache.py:
```diff
@@ -896,6 +896,14 @@ def compile(cls, graph, source_code, cuda):
+                if serialized_extern_kernel_nodes:
+                    assert (
+                        config.is_fbcode()
+                    ), "Only support serializing extern nodes in fbcode"
```
This test is needed. 
fbcode'ness is already tested in a few places. 
- V.graph.extern_kernel_nodes and V.graph.extern_node_serializer will be populated if only it's under fbcode. 
- serializer is used if only it's under fbcode
- serialized result is written to disk if only it's under fbcode. 

Unfortunately, these are scatter around in the code due to existing code structure.

SherlockNoMad commented on torch/_inductor/ir.py:
```diff
@@ -3690,8 +3701,42 @@ def find_device(tensor_args, example_output):
+        serializer = GraphModuleSerializer(None, None, None)
+        named_arguments = serializer.serialize_inputs(self.op_overload, args, kwargs)
+
+        # TODO: only single output is supported
```
Agree. Will work on it in latter PRs.

SherlockNoMad committed:
[AOTInductor] Proxy Executor for Extern Fallback kernels (#107279)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/107279

This is a prototype for running extern fallback kernels with a host side proxy executor.

Sample of generated cpp wrapper call:
```
        at::Tensor buf0;  // output buffer
        void* tensor_args_var_0[] = {&arg0_1, &arg0_1, &arg1_1, &arg0_1, &arg1_1, &buf0};
        int64_t int_args_var_1[] = {81, 81, 7, 7, 7, 81};
        proxy_executor->call_function("buf0", int_args_var_1, tensor_args_var_0);
```

- In my current implementation, proxy executor interprets the raw pointers according to the ops schema.
This assumes that custom op MUST have a valid schema registered to Dispatcher. (I would like to validate this assumption)
- I am using callboxed() API of the custom kernels. This is inevitable, as we wish to have a single call_function API for all possible custom kernels.

- These are all the input argument types I have support so far.
       union Argument {
         # Bool value does not matter
         1: bool asNone;
         2: TensorArgument asTensor;
         3: list<TensorArgument> asTensors;
         5: i64 asInt;
         7: list<i64> asInts;
         8: double asFloat;
         9: list<double> asFloats;
         10: string asString;
         10.5: list<string> asStrings;
         11: SymIntArgument asSymInt;
         12: list<SymIntArgument> asSymInts;
         13: ScalarType asScalarType;
         14: MemoryFormat asMemoryFormat;
         15: Layout asLayout;
         16: Device asDevice;
         17: bool asBool;
         18: list<bool> asBools;
       }

- Need a policy for handling unpopulated argument with default values. Here are the options, and it has BC  implications.
1. requires exported fx graph to explicitly populate default values, if users doesn't specify.
2. requires cpp wrapper to explicitly populate default values, if fx graph doesn't specify.
3. Proxy executor look up from opSchema for default values.

Test Plan:
frontend:
buck2 run mode/dev-sand mode/inplace -c fbcode.enable_gpu_sections=True sigmoid/frontend:export_main

test:
 buck2 run mode/dev-sand //deeplearning/aot_inductor/test:test_custom_ops

backend:
buck2 run mode/dev-nosan //deeplearning/aot_inductor/fb:main

Reviewed By: ezyang

Differential Revision: D48006843

fbshipit-source-id: fe394f24399a21f2cbfc5b343f39350d838b8fc6

izaitsevfb commented:
Adding @diff-train-skip-merge to (hopefully) unblock ShipIt.

SherlockNoMad commented:
Relanding with another diff and PR. aborting this.
